pkg/inclusion/sub_tree_root.go

```diff
@@ -0,0 +1,125 @@
+package inclusion
+
+import (
+	"fmt"
+
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+	"github.com/tendermint/tendermint/pkg/da"
+	"github.com/tendermint/tendermint/pkg/wrapper"
+)
+
+// TODO optimize https://github.com/celestiaorg/nmt/blob/e679661c6776d8a694f4f7c423c2e2eccb6c5aaa/subrootpaths.go#L15-L28
+
+// subTreeRootCacher keep track of all the inner nodes of an nmt using a simple
+// map. Note: this cacher does not cache individual leaves or their hashes, only
+// inner nodes.
+type subTreeRootCacher struct {
+	cache map[string][2]string
+}
+
+func newSubTreeRootCacher() *subTreeRootCacher {
+	return &subTreeRootCacher{cache: make(map[string][2]string)}
+}
+
+// Visit fullfills the nmt.VisitorNode function definition. It stores each inner
+// node in a simple map, which can later be used to walk the tree. This function
+// is called by the nmt when calculating the root.
+func (strc *subTreeRootCacher) Visit(hash []byte, children ...[]byte) {
+	switch len(children) {
+	case 2:
+		strc.cache[string(hash)] = [2]string{string(children[0]), string(children[1])}
+	case 1:
+		return
+	default:
+		panic("unexpected visit")
+	}
+}
+
+// walk recursively traverses the subTreeRootCacher's internal tree by using the
+// provided sub tree root and path. The provided path should be a []bool, false
+// indicating that the first child node (right most node) should be used to find
```

Comment:
I think these are flipped. The implementation walks left if false

```suggestion
// indicating that the first child node (left most node) should be used to find
```